### PR TITLE
Improvement to the `stored_size` code.

### PR DESCRIPTION
**Pull request by MathieuDutSik**
## Motivation

The merging of the PR #1287 was actually done a little bit prematurely, so this PR actually
do after merge corrections.

## Proposal

The following is done:
* Add documentation for the subtle `|| self.was_cleared` for the `total_size`.
* Do the same trick for the hash.
* Replace the `ByteMapView<SizeData>` by `ByteMapView<u32>` which takes less space.
* Add the check of overflow for the `add_assign`.

## Test Plan

The CI does it.

## Release Plan

That should not change anything.

## Links

- [reviewer checklist](https://github.com/linera-io/linera-protocol/blob/main/CONTRIBUTING.md#reviewer-checklist)

### CODE REVIEW

File: linera-views/src/collection_view.rs [hunk 1]
```diff
@@ -119,14 +118,19 @@ where
             }
         }
         let hash = *self.hash.get_mut();
-        if self.stored_hash != hash {
+        // In tne admittedly rare scenarion that we do a clear
```
Comment (ma2bd) [inline linera-views/src/collection_view.rs:121]: fix spelling: `tne` `scenarion`

Comment (ma2bd) [inline linera-views/src/collection_view.rs:121]: fix spelling: `tne` `scenarion`

`scenario that` -> `scenario where` (I think?)

File: linera-views/src/key_value_store_view.rs [hunk 2]
```diff
@@ -168,14 +175,21 @@ where
         }
         self.sizes.flush(batch)?;
         let hash = *self.hash.get_mut();
-        if self.stored_hash != hash {
+        // In tne admittedly rare scenarion that we do a clear
```
Comment (ma2bd) [inline linera-views/src/key_value_store_view.rs:178]: same

File: linera-views/src/collection_view.rs (hunk 1 above)
Comment (ma2bd) [reply linera-views/src/collection_view.rs:121]: nit: remove "admittedly rare" (in technical documentation, the shorter the better)

File: linera-views/src/key_value_store_view.rs [hunk 3]
```diff
@@ -605,8 +619,12 @@ where
             match operation {
                 WriteOperation::Delete { key } => {
                     ensure!(key.len() <= max_key_size, ViewError::KeyTooLong);
-                    if let Some(size) = self.sizes.get(&key).await? {
-                        self.total_size.sub_assign(size);
+                    if let Some(value) = self.sizes.get(&key).await? {
+                        let single_size = SizeData {
+                            key: key.len() as u32,
```
Comment (ma2bd) [inline linera-views/src/key_value_store_view.rs:624]: Do we need to deal with the overflow?

File: linera-views/src/key_value_store_view.rs [hunk 4]
```diff
@@ -605,8 +619,12 @@ where
             match operation {
                 WriteOperation::Delete { key } => {
                     ensure!(key.len() <= max_key_size, ViewError::KeyTooLong);
-                    if let Some(size) = self.sizes.get(&key).await? {
-                        self.total_size.sub_assign(size);
+                    if let Some(value) = self.sizes.get(&key).await? {
+                        let single_size = SizeData {
```
Comment (ma2bd) [inline linera-views/src/key_value_store_view.rs:623]: nit: `let size` or `let entry_size` ?

File: linera-views/src/key_value_store_view.rs (hunk 3 above)
Comment (MathieuDutSik) [reply linera-views/src/key_value_store_view.rs:624]: u32 corresponds to 4G.
So, it is a very big size, but not something that we cannot imagine.

File: linera-views/src/key_value_store_view.rs (hunk 4 above)
Comment (MathieuDutSik) [reply linera-views/src/key_value_store_view.rs:623]: The `entry_size` has my preference.

File: linera-views/src/collection_view.rs (hunk 1 above)
Comment (MathieuDutSik) [reply linera-views/src/collection_view.rs:121]: Yes, for `scenario where`.

Comment (MathieuDutSik) [reply linera-views/src/collection_view.rs:121]: Yes.

File: linera-views/src/key_value_store_view.rs [hunk 5]
```diff
@@ -63,9 +63,16 @@ pub struct SizeData {
...
+            .key
+            .checked_add(size.key)
+            .ok_or(ViewError::ArithmeticError(ArithmeticError::Overflow))?;
+        self.value = self
+            .value
+            .checked_add(size.value)
+            .ok_or(ViewError::ArithmeticError(ArithmeticError::Overflow))?;
+        Ok(())
     }
     /// Subtract a size to the existing SizeData
```
Comment (afck) [inline linera-views/src/key_value_store_view.rs:77]:
```suggestion

    /// Subtract a size to the existing SizeData
```

File: linera-views/src/key_value_store_view.rs (hunk 3 above)
Comment (afck) [reply linera-views/src/key_value_store_view.rs:624]: So it should probably be something like:

```suggestion
                            key: u32::try_from(key.len()).map_err(|_| ArithmeticError::Overflow)?,
```

Comment (ma2bd) [reply linera-views/src/key_value_store_view.rs:624]: It surely doesn't hurt to use the safe APIs. We also have a notion of max key size that might prevent this (does it?)

**Review by afck — APPROVED** [review]

**Review by ma2bd — CHANGES_REQUESTED** [review]
I'm going to be the bad guy here and point out that the test plan is incomplete (doesn't cover the fixed bug)

**Pull request merged**